guacamole-ext/src/main/java/org/apache/guacamole/net/auth/Connectable.java

```diff
@@ -54,8 +97,13 @@
      *     If an error occurs while connecting to guacd, or if permission to
      *     connect is denied.
      */
-    public GuacamoleTunnel connect(GuacamoleClientInformation info,
-            Map<String, String> tokens) throws GuacamoleException;
+    default GuacamoleTunnel connect(GuacamoleClientInformation info,
+            Map<String, String> tokens) throws GuacamoleException {
+
+        // Allow old implementations of Connectable to continue to work
+        return this.connect(info);
```

Comment:
Is there a reason supplying a default of this particular instance of this method is necessary or valuable?  Won't the old implementations already be calling the other default method, without the `tokens` parameter?  And doesn't this introduce the possibility that an implementing class could end up recursively calling these parameters (if they don't actually implement either of them)?  Or will that get caught at compile time?

Reply:
> Is there a reason supplying a default of this particular instance of this method is necessary or valuable? Won't the old implementations already be calling the other default method, without the `tokens` parameter?

Some sort of default is necessary to bridge the implementations, as it's the webapp that ultimately calls `connect()`, and in this case will be calling the version which has `tokens`.

In the general case, defaults are needed to bridge in both directions to allow extensions of different versions to interoperate. An older extension (compatible with 1.0.0) which decorates the connections of other extensions will need to be able to invoke `connect()`, even though it will be calling the older version. A newer extension and the webapp need the same ability but in the other direction.

> And doesn't this introduce the possibility that an implementing class could end up recursively calling these parameters (if they don't actually implement either of them)?

Yes, and that's definitely a bad thing. I think I may have a route around this.

Reply:
OK - I'm rearranging things a bit such that the interface defined within guacamole-ext is still strictly as defined before: one `connect()` definition which accepts `tokens`.

Within the webapp, I'm adding another version of `Connectable` in the same package which defines both `connect()` versions as well as defaults, and thus an ABI-compatible redefinition of both versions of the interface. Because of the way servlet containers handle classes bundled with the webapp vs. within dependencies, the webapp version of `Connectable` takes priority.